llvm/lib/Target/RISCV/RISCVInstrInfo.cpp

```diff
@@ -1446,6 +1446,196 @@ RISCVInstrInfo::optimizeSelect(MachineInstr &MI,
+  case RISCVCC::COND_EQ:
+  case RISCVCC::COND_NE: {
+    Register XorReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    BuildMI(MBB, MI, DL, get(RISCV::XOR), XorReg).addReg(LHSReg).addReg(RHSReg);
+    if (CC == RISCVCC::COND_EQ) {
+      BuildMI(MBB, MI, DL, get(RISCV::SLTIU), DstReg).addReg(XorReg).addImm(1);
+      return;
+    } else {
+      BuildMI(MBB, MI, DL, get(RISCV::SLTU), DstReg)
+          .addReg(RISCV::X0)
+          .addReg(XorReg);
+      return;
+    }
+  }
+  case RISCVCC::COND_GE:
+  case RISCVCC::COND_GEU: {
+    Register NotCCReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    BuildMI(MBB, MI, DL, get(CC == RISCVCC::COND_GE ? RISCV::SLT : RISCV::SLTU),
+            NotCCReg)
+        .addReg(LHSReg)
+        .addReg(RHSReg);
+    BuildMI(MBB, MI, DL, get(RISCV::XORI), DstReg).addReg(NotCCReg).addImm(1);
+    return;
+  }
+  }
+}
+
+void RISCVInstrInfo::insertSelect(MachineBasicBlock &MBB,
+                                  MachineBasicBlock::iterator MI,
+                                  const DebugLoc &DL, Register DstReg,
+                                  ArrayRef<MachineOperand> Cond,
+                                  Register TrueReg, Register FalseReg) const {
+  MachineFunction &MF = *MI->getParent()->getParent();
+  const RISCVSubtarget &ST = MF.getSubtarget<RISCVSubtarget>();
+  const TargetRegisterInfo &TRI = *ST.getRegisterInfo();
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  Register CCReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+
+  unsigned CC = Cond[0].getImm();
+  Register LHSReg = Cond[1].getReg();
+  Register RHSReg = Cond[2].getReg();
+
+  unsigned CondZeroEqzOpc =
+      ST.hasVendorXVentanaCondOps() ? RISCV::VT_MASKC : RISCV::CZERO_EQZ;
+  unsigned CondZeroNezOpc =
+      ST.hasVendorXVentanaCondOps() ? RISCV::VT_MASKCN : RISCV::CZERO_NEZ;
+
+  const TargetRegisterClass *DstRC = MRI.getRegClass(DstReg);
+  const TargetRegisterClass *CommonRC =
+      TRI.getCommonSubClass(DstRC, &RISCV::GPRRegClass);
+  bool NeedsRCCopies = (CommonRC != DstRC) && (CommonRC != &RISCV::GPRRegClass);
+
+  Register CondZeroEqzReg = TrueReg;
+  Register CondZeroNezReg = FalseReg;
+  if (NeedsRCCopies) {
+    CondZeroEqzReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    BuildMI(MBB, MI, DL, get(TargetOpcode::COPY), CondZeroEqzReg)
+        .addReg(TrueReg);
+    CondZeroNezReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    BuildMI(MBB, MI, DL, get(TargetOpcode::COPY), CondZeroNezReg)
+        .addReg(FalseReg);
+  }
+  if (CC == RISCVCC::COND_GE || CC == RISCVCC::COND_GEU) {
+    CC = (CC == RISCVCC::COND_GE) ? RISCVCC::COND_LT : RISCVCC::COND_LTU;
+    std::swap(CondZeroEqzReg, CondZeroNezReg);
+  }
+  insertICmp(MBB, MI, DL, CCReg, CC, LHSReg, RHSReg);
+
+  Register TrueValOrZeroReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+  Register FalseValOrZeroReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+  BuildMI(MBB, MI, DL, get(CondZeroEqzOpc), TrueValOrZeroReg)
```

Comment:
We may need to implement all the patterns in SDAG ISel here to reduce some instructions.

Reply:
Or, we implement these ISel patterns in a MIR pass, and then run the pass after early if-conversion pass.

Reply:
These patterns apply when one of the sides is zero. In our case we are looking at select between the results of two basic blocks. How often will it be that the result of a basic block is just zero?

Reply:
I have no data to show how common these cases will be, but we have do a lot of peepholes in `RISCVInstrInfoZicond.td` and `RISCVTargetLowering::lowerSELECT` to reduce the comparison and Zicond instructions.
After if-conversion, we need to do the same peepholes (which means we may not need czero.eqz+czero.nez+or to simulate `select`).

Reply:
I see what you mean. This seems like a lot of work and code duplication to reimplement all of these here.  Here's a list of all peepholes that we do in `lowerSELECT`:

```
(select c, t, 0) -> (czero_eqz t, c)
(select c, 0, f) -> (czero_nez f, c)
(select c, (and f, x), f) -> (or (and f, x), (czero_nez f, c))
(select c, t, (and t, x)) -> (or (czero_eqz t, c), (and t, x))
(select c, c1, c2) -> (add (czero_nez c2 - c1, c), c1)
(select c, c1, c2) -> (add (czero_eqz c1 - c2, c), c2)

(select c, -1, y) -> -c | y
(select c, y, -1) -> (c-1) | y
(select c, 0, y) -> (c-1) & y
(select c, y, 0) -> -c & y
select c, ~x, x --> xor -c, x
(select x, x, y) -> x | y
(select !x, x, y) -> x & y
(select x, y, x) -> x & y
(select !x, y, x) -> x | y
binOp ((select cond, x, c0), c1) -> select cond binOp(x, c1), binOp(c0, c1)
```
But perhaps we can address this later in a different MR, this MR is just the basic functionality.

Reply:
> This seems like a lot of work and code duplication to reimplement all of these here. 

Yes! So I said `Or, we implement these ISel patterns in a MIR pass and then run the pass after early if-conversion pass`. :-)

> But perhaps we can address this later in a different MR, this MR is just the basic functionality.

Agree. Can you make this PR non-draft? I ran llvm-testsuite and some internel tests and I haven't found any correctness issue, so I think it's ready for reviewing.

Reply:
This needs to be merged first: https://github.com/llvm/llvm-project/pull/95877
This MR uses a hack to bypass that.

Also I need to add some tests too.

Reply:
It's OK to be stacked on other PRs, make it non-draft so that more reviewers can see it.